### PR TITLE
feat(router): Allow Route.redirectTo to be a function which returns a…

### LINKED ISSUE #13373: Route redirectTo as function / class
**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```
**What is the motivation / use case for changing the behavior?**

Currently, in a `Route` definition, `redirectTo` can only be a string.In order to provide higher flexibility, it could also be a object, like the `Rezolve`.
The advantages related to the control we can have over the route parameters and the queryParams, or even dynamic redirection

* **Language:** Typescript

### LINKED ISSUE #28661: Preserve query params in redirectTo for absolute paths
As discussed in #13315, query params are only preserved when redirecting if the path is relative. This is maddeningly subtle, and I can't imagine that it's an intentional feature. I think query params should always be preserved across redirectTo-based redirects, even those with absolute paths.

### PR DESCRIPTION
**Pull request by atscott**
… string or UrlTree

This commit updates the logic around `Route.redirectTo` to enable using a function to create the redirect. This function can return a string, ands acts the same as previous string redirects, or a `UrlTree`, which will act as an absolute redirect.

To be useful, the redirect function needs access to the params and data. Today, developers can access these in their redirect strings, for example `{path: ':id', redirectTo: '/user/:id'}`. Unfortunately, developers only have access to params and data on the _current route_ today in the redirect strings. The params and data in the `RedirectFn` give developers access to anything from the matched parent routes as well. This is done as the same way as param and data aggregation later on (https://github.com/angular/angular/blob/897f014785578d87bc655ea6ae9e113653960f50/packages/router/src/router_state.ts#L236-L278). In order to accomplish this, we inherit params and data while matching, after the `ActivatedRouteSnapshot` is created for the matched route rather than waiting until the end.

The `RedirectFunction` does not return the full
`ActivatedRouteSnapshot` interface. Some things are not accurately known at the route matching phase. For example, resolvers are not run until later, so any resolved title would not be populated. The same goes for lazy loaded components. The is also true for all the snapshots up to the root, so properties that include parents (root, parent, pathFromRoot) are also excluded. And naturally, the full route matching hasn't yet happened so firstChild and children are not available either.

fixes #13373
resolves #28661 (though not for the redirect string - you would return a `UrlTree` from the function)

### CODE REVIEW

**Comment by atscott** [comment]
Updating target to `major`. This will likely need to be considered a breaking change. Expanding the type on the `redirectTo` results in build failures because places looking at the route config and expecting it to be a string will not work without a migration.

File: packages/router/src/recognize.ts [hunk 1]
```diff
@@ -407,6 +425,9 @@ export class Recognizer {
               route,
               getResolve(route),
             );
+            const i = getInherited(snapshot, parentRoute, this.paramsInheritanceStrategy);
```
Comment (AndrewKushnir) [inline packages/router/src/recognize.ts:428]: nit: I'd propose using a full name:
```suggestion
            const inherited = getInherited(snapshot, parentRoute, this.paramsInheritanceStrategy);
```
(we may need to pass it somewhere or use in a different part of the function later, so it'd be easier to refactor later)

File: packages/router/src/models.ts [hunk 2]
```diff
@@ -225,6 +225,13 @@ export type LoadChildren = LoadChildrenCallback;
  */
 export type QueryParamsHandling = 'merge' | 'preserve' | '';
 
+export type RedirectFunction = (
```
Comment (AndrewKushnir) [inline packages/router/src/models.ts:228]: This function should be exposed in the public API as well?

**Review by AndrewKushnir — APPROVED** [review]

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

**Comment by atscott** [comment]
TESTED=https://fusion2.corp.google.com/presubmit/582101259/OCL:582101259:BASE:615787806:1710430124292:2547af3b;groups=PossiblyAlreadyFailing/targets (1 pre-existing failure)

**Comment by atscott** [comment]
This PR was merged into the repository by commit 2b802587f27186baa493c1dd01f42d568b652f38.

**Pull request closed without merging**